File: mpi/comm.c
/*
comm.c (MPI)

MPI routines
*/

#ifdef _MPI
#include <mpi.h>
#endif

#include "ost.h"
#include "ost_prototype.h"

// initialize
void mpi_init(int argc, char **argv)
{
#ifdef _MPI
	MPI_Init(&argc, &argv);
	MPI_Comm_size(MPI_COMM_WORLD, &commSize);
	MPI_Comm_rank(MPI_COMM_WORLD, &commRank);
#else
	commSize = 1;
	commRank = 0;
	argc = argc;	// dummy
	argv = argv;	// dummy
#endif
}


// close
void mpi_close(void)
{
#ifdef _MPI
	MPI_Finalize();
#endif
}


// check error code
// mode = 0/1 : Bcast/Allreduce
void comm_check(int ierr, int mode, int prompt)
{
#ifdef _MPI
	if (commSize > 1) {
		if (mode == 0) {
			MPI_Bcast(&ierr, 1, MPI_INT, 0, MPI_COMM_WORLD);
		}
		else {
			int g_ierr;
			MPI_Allreduce(&ierr, &g_ierr, 1, MPI_INT, MPI_LOR, MPI_COMM_WORLD);
			ierr = g_ierr;
		}
	}
	if (ierr) {
		MPI_Finalize();
	}
#endif
	mode = 0;  // dummy
	if (ierr) {
		if (prompt && (commRank == 0)) {
			fflush(stdout);
			getchar();
		}
		exit(0);
	}
}


// get cpu time [sec]
double comm_cputime(void)
{
#ifdef _MPI
	MPI_Barrier(MPI_COMM_WORLD);
	return MPI_Wtime();
#else
#ifdef _WIN32
	return (double)clock() / CLOCKS_PER_SEC;
#else
	struct timespec ts;
	clock_gettime(CLOCK_REALTIME, &ts);
	return (ts.tv_sec + (ts.tv_nsec * 1e-9));
#endif  // _WIN32
#endif  // MPI
}


// gather string
void comm_string(const char *lstr, char *gstr)
{
#ifdef _MPI
	char buff[BUFSIZ];
	if (commRank == 0) {
		MPI_Status status;
		strcpy(gstr, lstr);
		for (int i = 1; i < commSize; i++) {
			MPI_Recv(buff, BUFSIZ, MPI_CHAR, i, 0, MPI_COMM_WORLD, &status);
			strcat(gstr, "\n");
			strcat(gstr, buff);
		}
	}
	else {
		strcpy(buff, lstr);
		MPI_Send(buff, BUFSIZ, MPI_CHAR, 0, 0, MPI_COMM_WORLD);
	}
#else
	strcpy(gstr, lstr);
#endif
}


// broadcast input data
void comm_broadcast(void)
{
#ifdef _MPI
	const int maxvolt = sizeof(Volt) / sizeof(double);  // =256
	const int maxepsr = sizeof(Epsr) / sizeof(real_t);  // =256

	int    *i_buf = NULL;
	double *d_buf = NULL;
	int i_num = 0;
	int d_num = 0;

	// variables to buffers (root only)

	if (commRank == 0) {
		// number of data
		i_num = 8 + (3 * NGeom);
		d_num = 2 + (Nx + 1) + (Ny + 1) + (Nz + 1)
		      + (1 * maxvolt) + (1 * maxepsr) + (8 * NGeom);

		// alloc
		i_buf =    (int *)malloc(i_num * sizeof(int));
		d_buf = (double *)malloc(d_num * sizeof(double));

		int i_id = 0;
		int d_id = 0;

		i_buf[i_id++] = Nx;
		i_buf[i_id++] = Ny;
		i_buf[i_id++] = Nz;
		i_buf[i_id++] = NVolt;
		i_buf[i_id++] = NEpsr;
		i_buf[i_id++] = NGeom;

		i_buf[i_id++] = Solver.maxiter;
		i_buf[i_id++] = Solver.nout;
		d_buf[d_id++] = Solver.omega;
		d_buf[d_id++] = Solver.converg;

		for (int i = 0; i <= Nx; i++) {
			d_buf[d_id++] = Xn[i];
		}

		for (int j = 0; j <= Ny; j++) {
			d_buf[d_id++] = Yn[j];
		}

		for (int k = 0; k <= Nz; k++) {
			d_buf[d_id++] = Zn[k];
		}

		for (int n = 0; n < maxvolt; n++) {
			d_buf[d_id++] = Volt[n];
		}

		for (int n = 0; n < maxepsr; n++) {
			d_buf[d_id++] = Epsr[n];
		}

		for (int n = 0; n < NGeom; n++) {
			i_buf[i_id++] = Geom[n].type;
			i_buf[i_id++] = Geom[n].shape;
			i_buf[i_id++] = Geom[n].pid;
			for (int m = 0; m < 8; m++) {
				d_buf[d_id++] = Geom[n].g[m];
			}
		}

		// check
		assert(i_id == i_num);
		assert(d_id == d_num);
	}

	// broadcast (root to non-root)

	MPI_Bcast(&i_num, 1, MPI_INT, 0, MPI_COMM_WORLD);
	MPI_Bcast(&d_num, 1, MPI_INT, 0, MPI_COMM_WORLD);

	// alloc
	if (commRank > 0) {
		i_buf =    (int *)malloc(i_num * sizeof(int));
		d_buf = (double *)malloc(d_num * sizeof(double));
	}

	MPI_Bcast(i_buf, i_num, MPI_INT,    0, MPI_COMM_WORLD);
	MPI_Bcast(d_buf, d_num, MPI_DOUBLE, 0, MPI_COMM_WORLD);

	if (commRank > 0) {
		int i_id = 0;
		int d_id = 0;

		Nx             = i_buf[i_id++];
		Ny             = i_buf[i_id++];
		Nz             = i_buf[i_id++];
		NVolt          = i_buf[i_id++];
		NEpsr          = i_buf[i_id++];
		NGeom          = i_buf[i_id++];

		Solver.maxiter = i_buf[i_id++];
		Solver.nout    = i_buf[i_id++];
		Solver.omega   = d_buf[d_id++];
		Solver.converg = d_buf[d_id++];

		Xn = (double *)malloc((Nx + 1) * sizeof(double));
		for (int i = 0; i <= Nx; i++) {
			Xn[i] = d_buf[d_id++];
		}

		Yn = (double *)malloc((Ny + 1) * sizeof(double));
		for (int j = 0; j <= Ny; j++) {
			Yn[j] = d_buf[d_id++];
		}

		Zn = (double *)malloc((Nz + 1) * sizeof(double));
		for (int k = 0; k <= Nz; k++) {
			Zn[k] = d_buf[d_id++];
		}

		for (int n = 0; n < maxvolt; n++) {
			Volt[n] = d_buf[d_id++];
		}

		for (int n = 0; n < maxepsr; n++) {
			Epsr[n] = (real_t)d_buf[d_id++];
		}

		Geom = (geometry_t *)malloc(NGeom * sizeof(geometry_t));
		for (int n = 0; n < NGeom; n++) {
			Geom[n].type  =       i_buf[i_id++];
			Geom[n].shape =       i_buf[i_id++];
			Geom[n].pid   = (id_t)i_buf[i_id++];
			for (int m = 0; m < 8; m++) {
				Geom[n].g[m] = d_buf[d_id++];
			}
		}

		// check
		assert(i_id == i_num);
		assert(d_id == d_num);
	}

	// free
	free(i_buf);
	free(d_buf);
/*
	// debug
	printf("%d %d\n", commSize, commRank);
	printf("%d %d %d\n", Nx, Ny, Nz);
	//printf("%d %d %g %g\n", Solver.maxiter, Solver.nout, Solver.omega, Solver.converg);
	//for (int i = 0; i <= Nx; i++) printf("%d %f\n", i, Xn[i]);
	//for (int j = 0; j <= Ny; j++) printf("%d %f\n", j, Yn[j]);
	//for (int k = 0; k <= Nz; k++) printf("%d %f\n", k, Zn[k]);
	//printf("%d %d\n", NVolt, NEpsr);
	//for (int n = 0; n < NVolt; n++) printf("%d %f\n", n, Volt[n]);
	//for (int n = 0; n < NEpsr; n++) printf("%d %f\n", n, Epsr[n]);
	//printf("%d\n", NGeom);
	//for (int n = 0; n < NGeom; n++) printf("%d %d %d %d %f %f %f %f %f %f\n", n, Geom[n].type, Geom[n].shape, Geom[n].pid, Geom[n].g[0], Geom[n].g[1], Geom[n].g[2], Geom[n].g[3], Geom[n].g[4], Geom[n].g[5]);
	fflush(stdout);
*/
#endif
}


// sum reduction (scalar)
real_t comm_sum(real_t l_sum)
{
	real_t sum;
#ifdef _MPI
	MPI_Allreduce(&l_sum, &sum, 1, MPI_REAL_T, MPI_SUM, MPI_COMM_WORLD);
#else
	sum = l_sum;
#endif
	return sum;
}


// sum reduction (vector)
void comm_reduction(void)
{
	for (int n = 0; n < NVolt; n++) {
		Echar[n] = comm_sum((real_t)Echar[n]);
	}
}


// gather
void comm_gather(void)
{
#ifdef _MPI
	int64_t g_n, n;
	MPI_Status status;
	int isend[11], irecv[11];
	const int tag = 0;

	// root : self copy to global array
	if (commRank == 0) {
		// save local index
		const int imin = iMin;
		const int imax = iMax;
		const int jmin = jMin;
		const int jmax = jMax;
		const int kmin = kMin;
		const int kmax = kMax;
		const int64_t ni = Ni;
		const int64_t nj = Nj;
		const int64_t nk = Nk;
		const int64_t n0 = N0;
		//const int64_t nn = NN;

		// new global index
		setupsize(1, 1, 1, 0);

		// alloc global array
		g_V      = (real_t *)malloc(NN * sizeof(real_t));
		g_idVolt =   (id_t *)malloc(NN * sizeof(id_t));
		g_idEpsr =   (id_t *)malloc(NN * sizeof(id_t));

		// self copy
		for (int i = imin; i <= imax; i++) {
		for (int j = jmin; j <= jmax; j++) {
		for (int k = kmin; k <= kmax; k++) {
			g_n = (Ni * i) + (Nj * j) + (Nk * k) + N0;
			n   = (ni * i) + (nj * j) + (nk * k) + n0;
			g_V[g_n]      = V[n];
			g_idVolt[g_n] = idVolt[n];
			g_idEpsr[g_n] = idEpsr[n];
		}
		}
		}
	}

	// root : receive
	if (commRank == 0) {
		for (int irank = 1; irank < commSize; irank++) {
			// receive local index
			MPI_Recv(irecv, 11, MPI_INT, irank, tag, MPI_COMM_WORLD, &status);
			const int imin = irecv[0];
			const int imax = irecv[1];
			const int jmin = irecv[2];
			const int jmax = irecv[3];
			const int kmin = irecv[4];
			const int kmax = irecv[5];
			const int ni   = irecv[6];
			const int nj   = irecv[7];
			const int nk   = irecv[8];
			const int n0   = irecv[9];
			const int nn   = irecv[10];
			//printf("%d %d %d %d %d\n", commRank, imin, imax, n0, nn); fflush(stdout);

			// alloc receive buffer
			real_t *recv_v      = (real_t *)malloc(nn * sizeof(real_t));
			id_t   *recv_idvolt =   (id_t *)malloc(nn * sizeof(id_t));
			id_t   *recv_idepsr =   (id_t *)malloc(nn * sizeof(id_t));

			// receive and set
			MPI_Recv(recv_v,      nn, MPI_REAL_T, irank, tag, MPI_COMM_WORLD, &status);
			MPI_Recv(recv_idvolt, nn, MPI_ID_T,   irank, tag, MPI_COMM_WORLD, &status);
			MPI_Recv(recv_idepsr, nn, MPI_ID_T,   irank, tag, MPI_COMM_WORLD, &status);
			for (int i = imin; i <= imax; i++) {
			for (int j = jmin; j <= jmax; j++) {
			for (int k = kmin; k <= kmax; k++) {
				g_n = (Ni * i) + (Nj * j) + (Nk * k) + N0;
				n   = (ni * i) + (nj * j) + (nk * k) + n0;
				g_V[g_n]      = recv_v[n];
				g_idVolt[g_n] = recv_idvolt[n];
				g_idEpsr[g_n] = recv_idepsr[n];
			}
			}
			}
		}
	}
	else {
		// non-root : send to root
		// index
		isend[0]  = iMin;
		isend[1]  = iMax;
		isend[2]  = jMin;
		isend[3]  = jMax;
		isend[4]  = kMin;
		isend[5]  = kMax;
		isend[6]  = (int)Ni;
		isend[7]  = (int)Nj;
		isend[8]  = (int)Nk;
		isend[9]  = (int)N0;
		isend[10] = (int)NN;
		MPI_Send(isend, 11, MPI_INT, 0, tag, MPI_COMM_WORLD);

		// data
		const int nn = (int)NN;
		MPI_Send(V,      nn, MPI_REAL_T, 0, tag, MPI_COMM_WORLD);
		MPI_Send(idVolt, nn, MPI_ID_T,   0, tag, MPI_COMM_WORLD);
		MPI_Send(idEpsr, nn, MPI_ID_T,   0, tag, MPI_COMM_WORLD);
	}

	// copy pointer : global -> root
	if (commRank == 0) {
		V      = g_V;
		idVolt = g_idVolt;
		idEpsr = g_idEpsr;
	}
#endif
}

File: python/sol/comm.py
# -*- coding: utf-8 -*-
"""
comm.py (MPI)
"""

import numpy as np
from mpi4py import MPI
from numba import jit, prange
import sol.setup

# 入力データを全プロセスで共有する
def broadcast(comm_rank,
    Nx, Ny, Nz, Xn, Yn, Zn, fVolt, fEpsr, iGeometry, fGeometry, Parm):

    i_buf = None
    f_buf = None
    i_num = np.zeros(1, 'i4')
    f_num = np.zeros(1, 'i4')

    # 変数をバッファに格納する (root)
    if comm_rank == 0:
        # 配列の大きさ
        nvolt = len(fVolt)
        nepsr = len(fEpsr)
        ngeometry = iGeometry.shape[0]

        # データ数
        i_num[0] = 8 + (2 * ngeometry)
        f_num[0] = 2 + (Nx + 1) + (Ny + 1) + (Nz + 1) + nvolt + nepsr + (8 * ngeometry)

        # alloc
        i_buf = np.zeros(i_num[0], 'i4')
        f_buf = np.zeros(f_num[0], 'f8')

        i_id = 0
        f_id = 0

        i_buf[i_id] = Nx; i_id += 1
        i_buf[i_id] = Ny; i_id += 1
        i_buf[i_id] = Nz; i_id += 1

        i_buf[i_id] = nvolt; i_id += 1
        i_buf[i_id] = nepsr; i_id += 1
        i_buf[i_id] = ngeometry; i_id += 1

        f_buf[f_id] = Parm['solver'][0]; f_id += 1
        i_buf[i_id] = Parm['solver'][1]; i_id += 1
        i_buf[i_id] = Parm['solver'][2]; i_id += 1
        f_buf[f_id] = Parm['solver'][3]; f_id += 1

        for i in range(Nx + 1):
            f_buf[f_id] = Xn[i]; f_id += 1

        for j in range(Ny + 1):
            f_buf[f_id] = Yn[j]; f_id += 1

        for k in range(Nz + 1):
            f_buf[f_id] = Zn[k]; f_id += 1

        for n in range(nvolt):
            f_buf[f_id] = fVolt[n]; f_id += 1

        for n in range(nepsr):
            f_buf[f_id] = fEpsr[n]; f_id += 1

        for n in range(ngeometry):
            for m in range(2):
                i_buf[i_id] = iGeometry[n, m]; i_id += 1
            for m in range(8):
                f_buf[f_id] = fGeometry[n, m]; f_id += 1

        # check
        assert(i_id == i_num[0])
        assert(f_id == f_num[0])

    # broadcast (データ数、rootから非rootへ)
    MPI.COMM_WORLD.Bcast(i_num)
    MPI.COMM_WORLD.Bcast(f_num)
    #print(i_num, f_num)

    # alloc (非root)
    if comm_rank > 0:
        i_buf = np.zeros(i_num[0], 'i4')
        f_buf = np.zeros(f_num[0], 'f8')

    # broadcast (データ、rootから非rootへ)
    MPI.COMM_WORLD.Bcast(i_buf)
    MPI.COMM_WORLD.Bcast(f_buf)

    # 受信処理: 非root
    if comm_rank > 0:
        i_id = 0
        f_id = 0

        Nx        = i_buf[i_id]; i_id += 1
        Ny        = i_buf[i_id]; i_id += 1
        Nz        = i_buf[i_id]; i_id += 1
        nvolt     = i_buf[i_id]; i_id += 1
        nepsr     = i_buf[i_id]; i_id += 1
        ngeometry = i_buf[i_id]; i_id += 1
        #print(comm_rank, nvolt, nepsr, ngeometry)

        Parm['solver'] = [0] * 4
        Parm['solver'][0] = f_buf[f_id]; f_id += 1
        Parm['solver'][1] = i_buf[i_id]; i_id += 1
        Parm['solver'][2] = i_buf[i_id]; i_id += 1
        Parm['solver'][3] = f_buf[f_id]; f_id += 1

        Xn = np.zeros(Nx + 1, 'f8')
        for i in range(Nx + 1):
            Xn[i] = f_buf[f_id]; f_id += 1

        Yn = np.zeros(Ny + 1, 'f8')
        for j in range(Ny + 1):
            Yn[j] = f_buf[f_id]; f_id += 1

        Zn = np.zeros(Nz + 1, 'f8')
        for k in range(Nz + 1):
            Zn[k] = f_buf[f_id]; f_id += 1

        fVolt = np.zeros(nvolt, 'f8')
        for n in range(nvolt):
            fVolt[n] = f_buf[f_id]; f_id += 1

        fEpsr = np.zeros(nepsr, 'f8')
        for n in range(nepsr):
            fEpsr[n] = f_buf[f_id]; f_id += 1

        iGeometry = np.zeros((ngeometry, 2), 'i4')
        fGeometry = np.zeros((ngeometry, 8), 'f8')
        for n in range(ngeometry):
            for m in range(2):
                iGeometry[n, m] = i_buf[i_id]; i_id += 1
            for m in range(8):
                fGeometry[n, m] = f_buf[f_id]; f_id += 1

        # check
        assert(i_id == i_num[0])
        assert(f_id == f_num[0])

    # free
    i_buf = None
    f_buf = None

    #print(comm_rank, Nx, Ny, Nz)
    #print(comm_rank, Xn)
    #print(comm_rank, fVolt, fEpsr)
    #print(comm_rank, iGeometry, fGeometry)
    #print(comm_rank, Parm)

    return Nx, Ny, Nz, Xn, Yn, Zn, fVolt, fEpsr, iGeometry, fGeometry

# 全領域のデータをrootに集める
#@jit(cache=True, nopython=True, nogil=True, parallel=True)
def gather(Parm,
    V, idVolt, idEpsr,
    Nx, Ny, Nz, iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0, NN):

    comm_size = Parm['comm_size']
    comm_rank = Parm['comm_rank']
    #print(comm_size, comm_rank)

    # rootの前処理
    if comm_rank == 0:
        # 拡大前のindexを保存する
        imin = iMin
        imax = iMax
        jmin = jMin
        jmax = jMax
        kmin = kMin
        kmax = kMax
        ni = Ni
        nj = Nj
        nk = Nk
        n0 = N0
        #nn = NN

        # 全領域のindexを求める
        iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0, NN, Ipx, Ipy, Ipz \
        = sol.setup.getIndex(Nx, Ny, Nz, 1, 1, 1, 1, 1, 1, 0)
        #print(iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0, NN)

        # 全領域の配列を作成する
        #V      = np.resize(V,      NN)
        #idVolt = np.resize(idVolt, NN)
        #idEpsr = np.resize(idEpsr, NN)
        #print(f_dtype, i_dtype)
        g_V      = np.zeros(NN, Parm['f_dtype'])
        g_idVolt = np.zeros(NN, Parm['i_dtype'])
        g_idEpsr = np.zeros(NN, Parm['i_dtype'])

        # MPI用小領域のデータを全体配列にコピーする
        _copy3d3(g_V, g_idVolt, g_idEpsr, V, idVolt, idEpsr,
            imin, imax, jmin, jmax, kmin, kmax,
            Ni, Nj, Nk, N0,
            ni, nj, nk, n0)
        """
        for i in range(imin, imax + 1):
            for j in range(jmin, jmax + 1):
                for k in range(kmin, kmax + 1):
                    g_n = (Ni * i) + (Nj * j) + (Nk * k) + N0
                    n   = (ni * i) + (nj * j) + (nk * k) + n0
                    g_V[g_n]      = V[n]
                    g_idVolt[g_n] = idVolt[n]
                    g_idEpsr[g_n] = idEpsr[n]
        """

    # 送受信
    if comm_rank == 0:
        # root : 受信
        for rank in range(1, comm_size):
            # indexを受信する
            isize = np.zeros(11, 'i4')
            MPI.COMM_WORLD.Recv(isize, source=rank)
            [imin, imax, jmin, jmax, kmin, kmax, ni, nj, nk, n0, nn] = isize

            # 受信用配列を作成する
            recv_v      = np.zeros(nn, Parm['f_dtype'])
            recv_idvolt = np.zeros(nn, Parm['i_dtype'])
            recv_idepsr = np.zeros(nn, Parm['i_dtype'])

            # データを受信する
            MPI.COMM_WORLD.Recv(recv_v,      source=rank)
            MPI.COMM_WORLD.Recv(recv_idvolt, source=rank)
            MPI.COMM_WORLD.Recv(recv_idepsr, source=rank)
            #start = (Ni * isize[0]) + (Nj * (-1)) + (Nk * (-1)) + N0
            #count = (isize[1] - isize[0] + 1) * (Ny + 3) * (Nz + 3)
            #MPI.COMM_WORLD.Recv(     V[start: start + count], source=rank)
            #MPI.COMM_WORLD.Recv(idVolt[start: start + count], source=rank)
            #MPI.COMM_WORLD.Recv(idEpsr[start: start + count], source=rank)

            # 受信データを全体配列に代入する
            _copy3d3(g_V, g_idVolt, g_idEpsr, recv_v, recv_idvolt, recv_idepsr,
                imin, imax, jmin, jmax, kmin, kmax,
                Ni, Nj, Nk, N0,
                ni, nj, nk, n0)
            """
            for i in range(imin, imax + 1):
                for j in range(jmin, jmax + 1):
                    for k in range(kmin, kmax + 1):
                        g_n = (Ni * i) + (Nj * j) + (Nk * k) + N0
                        n   = (ni * i) + (nj * j) + (nk * k) + n0
                        g_V[g_n]      = recv_v[n]
                        g_idVolt[g_n] = recv_idvolt[n]
                        g_idEpsr[g_n] = recv_idepsr[n]
            """
    else:
        # 非root : 送信
        isize = np.array([iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0, NN], 'i4')
        MPI.COMM_WORLD.Send(isize, 0)
        MPI.COMM_WORLD.Send(     V, 0)
        MPI.COMM_WORLD.Send(idVolt, 0)
        MPI.COMM_WORLD.Send(idEpsr, 0)
        #start = (Ni * iMin) + (Nj * (-1)) + (Nk * (-1)) + N0
        #count = (iMax - iMin + 1) * (Ny + 3) * (Nz + 3)
        #MPI.COMM_WORLD.Send(     V[start: start + count], 0)
        #MPI.COMM_WORLD.Send(idVolt[start: start + count], 0)
        #MPI.COMM_WORLD.Send(idEpsr[start: start + count], 0)

    # 全体配列をrootの配列にコピーする
    if comm_rank == 0:
        V      = g_V.copy()
        idVolt = g_idVolt.copy()
        idEpsr = g_idEpsr.copy()

    return V, idVolt, idEpsr, \
        iMin, iMax, jMin, jMax, kMin, kMax, Ni, Nj, Nk, N0, NN

# (private) 3個の3D配列のコピー
@jit(cache=True, nopython=True, nogil=True, parallel=True)
def _copy3d3(a1, a2, a3, b1, b2, b3,
    imin, imax, jmin, jmax, kmin, kmax,
    ni_a, nj_a, nk_a, n0_a,
    ni_b, nj_b, nk_b, n0_b):
    for i in prange(imin, imax + 1):
        for j in range(jmin, jmax + 1):
            for k in range(kmin, kmax + 1):
                n_a = (ni_a * i) + (nj_a * j) + (nk_a * k) + n0_a
                n_b = (ni_b * i) + (nj_b * j) + (nk_b * k) + n0_b
                a1[n_a] = b1[n_b]
                a2[n_a] = b2[n_b]
                a3[n_a] = b3[n_b]

# 和(スカラー)
def sum_scalar(var):
    sendbuf = np.array([var], 'f8')
    recvbuf = np.zeros(1, 'f8')
    MPI.COMM_WORLD.Allreduce(sendbuf, recvbuf)

    return recvbuf[0]

# 和(ベクトル, rootのみに返す)
def sum_vector(var):
    #sendbuf = np.array(var, 'f8')
    #recvbuf = np.zeros(len(var), 'f8')
    #MPI.COMM_WORLD.Allreduce(sendbuf, recvbuf)
    #return recvbuf
    result = np.zeros(len(var), 'f8')
    MPI.COMM_WORLD.Reduce(var, result)
    return result
